### PR TITLE
Kernel documentation additions & rework

### LINKED ISSUE #24359: k_heap / sys_heap needs overview documentation
We've introduced new heap APIs to supersede k_mem_pool / sys_mem_pool, but have no overview documentation for it, just API doxygen.

I expect this file will need to be rewritten to reflect the new API set:

`doc/reference/kernel/memory/pools.rst`

### PR DESCRIPTION
**Pull request by andyross**
Dust off some older stuff from #19201 (which never merged because the timeout patch got stalled), update it for the review comments there and to conform to the API that merged.  Also adds a section on the new heap API.

Fixes: #24359

### CODE REVIEW

**Review by tyrantlizard77 — APPROVED** [review]
LGTM

You can mark this as Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/24359

Lintian checks found some typos, I presume you'll fix those.

**Comment by andyross** [comment]
I'm lost on that build failure.  The link from can_api.rst should go to doxygen output, and similar links to K_FOREVER in the new docs seem to be working.  The change doesn't remove a K_FOREVER target form the old docs...

File: doc/reference/kernel/timing/clocks.rst [hunk 1]
```diff
@@ -1,161 +1,377 @@
...
-*************
-
-.. doxygengroup:: clock_apis
-   :project: Zephyr
+    void my_wait_for_event(struct my_subsys *obj, k_timeout_t timeout_in_ms)
+    {
+        /* Compute the end time from the timeout */
+        u64_t end = z_timeout_end_calc(timeout_in_ms);
+
+            while (end < k_uptime_ticks()) {
```
Comment (mbolivar-nordic) [inline doc/reference/kernel/timing/clocks.rst:354]: Alignment looks funny. Is that just GitHub?

File: doc/reference/kernel/memory/heap.rst [hunk 2]
```diff
@@ -1,71 +1,121 @@
...
+combined with adjacent free blocks to prevent fragmentation.
+
+All metadata is stored at the beginning of the contiguous block of
+heap memory, including the variable-length list of bucket list heads
+(which depend on heap size).  The only external memory required is the
+:c:type:`struct sys_heap` structure itself.
+
+The ``sys_heap`` functions are unsynchronized.  Care must be taken by
+any users to prevent concurrent access.  Only one context may be
+inside one of the APIs functions at a time.
```
Comment (mbolivar-nordic) [inline doc/reference/kernel/memory/heap.rst:86]:
```suggestion
inside one of the API functions at a time.
```

File: doc/reference/kernel/memory/heap.rst [hunk 3]
```diff
@@ -94,7 +94,7 @@ the minimum bucket size for an allocation (the set of free blocks that
 unbounded list searches, at the expense of some fragmentation
 resistance.  This :c:option:`CONFIG_SYS_HEAP_ALLOC_LOOPS` value may be
 chosen by the user at build time, and defaults to a value of 3.
-   
+
```
Comment (mbolivar-nordic) [inline doc/reference/kernel/memory/heap.rst:97]: Did you mean to squash this one?

**Comment by mbolivar-nordic** [comment]
Thanks for these! They look great. I found some typos.

File: doc/reference/kernel/timing/clocks.rst [hunk 4]
```diff
@@ -1,161 +1,377 @@
...
+
+All these values are specified using a ``k_timeout_t`` value.  This is
+an opaque struct type that must be initialized using one of a family
+of ``K_TIMEOUT`` macros.  The most common, ``K_TIMEOUT_MS()``, defines
+a time in milliseconds after the current time (strictly: the time at
+which the kernel receives the timeout value).
+
+Other options for timeout initialization follow the unit conventions
+described above: ``K_TIMEOUT_US()``, ``K_TIMEOUT_TICKS()`` and
+``K_TIMEOUT_CYC()`` specify timeout values that will expire after
```
Comment (nashif) [inline doc/reference/kernel/timing/clocks.rst:107]: ditto for K_TIMEOUT_CYC

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by andyross** [comment]
Rebase, fix typos and some stale symbol names.  Seems like the CI failure had already been cleared.  Should be good to go, please rereview

**Review by nashif — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**